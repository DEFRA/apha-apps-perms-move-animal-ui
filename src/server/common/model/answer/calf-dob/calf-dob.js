import { DateAnswer } from '../date/date.js'

/** @import {DateConfig} from '../date/date.js' */

export class CalfDob extends DateAnswer {
  /** @type {DateConfig} */
  static config = {
    isPageHeading: true,
    validation: {
      missingDate: { message: 'Enter the oldest calfâ€™s date of birth' },
      missingDay: {
        message: 'Date of birth of the oldest calf must include a day'
      },
      missingMonth: {
        message: 'Date of birth of the oldest calf must include a month'
      },
      missingYear: {
        message: 'Date of birth of the oldest calf must include a year'
      },
      invalidDay: { message: 'Day of birth must be a real date' },
      invalidMonth: {
        message:
          'Month of birth of the oldest calf must be a number between 1 and 12'
      },
      invalidYear: { message: 'Year of birth must be a real date' },
      yearPattern: {
        message: 'Year of birth must include 4 numbers',
        pattern: /^\d{4}$/
      },
      invalidDate: { message: 'Day of birth must be a real date' },
      futureDate: {
        message: 'Date of birth of the oldest calf must be today or in the past'
      }
    }
  }
}
