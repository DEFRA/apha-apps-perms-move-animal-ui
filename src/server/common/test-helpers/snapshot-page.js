import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
