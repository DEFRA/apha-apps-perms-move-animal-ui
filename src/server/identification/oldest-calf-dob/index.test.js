import { CalfDob } from '../../common/model/answer/calf-dob/calf-dob.js'
import { describePageSnapshot } from '../../common/test-helpers/snapshot-page.js'
import { oldestCalfDobPage, OldestCalfDobPage } from './index.js'

const sectionKey = 'identification'
const question =
  'What is the date of birth of the oldest calf under 42 days old?'
const questionKey = 'oldestCalfDob'
const view = 'common/model/page/question-page.njk'
const pageUrl = '/identification/oldest-calf-dob'

describe('OldestCalfDobPage', () => {
  const page = new OldestCalfDobPage()

  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct view', () => {
    expect(page.view).toBe(view)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(CalfDob)
  })

  it('nextPage should return how to minimise contamination page when answer is "yes"', () => {
    // const answer = new CalvesUnder42DaysOldAnswer({
    //   calvesUnder42DaysOld: 'yes'
    // })
    // const nextPage = page.nextPage(answer)
    // expect(nextPage).toBe(oldestCalfDobPage)
  })

  it('nextPage should return people disinfection page when answer is "no"', () => {
    // const answer = new CalvesUnder42DaysOldAnswer({
    //   calvesUnder42DaysOld: 'no'
    // })
    // const nextPage = page.nextPage(answer)
    // expect(nextPage).toBe(testingDatesPage)
  })

  it('should export page', () => {
    expect(oldestCalfDobPage).toBeInstanceOf(OldestCalfDobPage)
  })

  describePageSnapshot({
    describes: 'oldestCalfDobPage.content',
    it: 'should render expected response and content',
    pageUrl
  })
})
