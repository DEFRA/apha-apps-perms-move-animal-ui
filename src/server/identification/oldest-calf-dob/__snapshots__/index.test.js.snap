// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OldestCalfDobPage oldestCalfDobPage.content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset" role="group">
    <legend class="govuk-fieldset__legend govuk-label--l">
      <h1 class="govuk-fieldset__heading">
        What is the date of birth of the oldest calf under 42 days old?
      </h1>
    </legend>
    <div class="govuk-date-input" id="date">
      <div class="govuk-date-input__item">
        <div class="govuk-form-group">
          <label class="govuk-label govuk-date-input__label" for="date-day">
            Day
          </label>
          <input class="govuk-input govuk-date-input__input govuk-input--width-2" id="date-day" name="day" type="text" inputmode="numeric">
        </div>
      </div>
      <div class="govuk-date-input__item">
        <div class="govuk-form-group">
          <label class="govuk-label govuk-date-input__label" for="date-month">
            Month
          </label>
          <input class="govuk-input govuk-date-input__input govuk-input--width-2" id="date-month" name="month" type="text" inputmode="numeric">
        </div>
      </div>
      <div class="govuk-date-input__item">
        <div class="govuk-form-group">
          <label class="govuk-label govuk-date-input__label" for="date-year">
            Year
          </label>
          <input class="govuk-input govuk-date-input__input govuk-input--width-4" id="date-year" name="year" type="text" inputmode="numeric">
        </div>
      </div>
    </div>
  </fieldset>
</div>

            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
